include/envoy/router/router.h

```diff
@@ -258,6 +258,13 @@ class VirtualHost {
    * @return const Config& the RouteConfiguration that owns this virtual host.
    */
   virtual const Config& routeConfig() const PURE;
+
+  /**
+   * @return const Protobuf::Message* the per-filter config for the given
+   * filter name configured for this virtual host. If none is present,
+   * nullptr is returned.
+   */
+  virtual const Protobuf::Message* perFilterConfig(const std::string& name) const PURE;
```

Comment:
This suggestion is trying to address this issue:  https://github.com/envoyproxy/envoy/issues/3008
We can do it in a separate PR.  But it is so simple, you can also put in in this PR too.

In these VHost/Route/Cluster,   add another function

// If perFitlerConfig exists for this filter, get its config hash value to detect if config is updated.
// return 0 if no config for the filter.
size_t perFitlerConfigHash(const std::string& name) const PURE;

Reply:
I don't quite understand how this will help you. In the filter are you going to check this on a per-route basis? What you really want is another callback which allows a filter to update global structures on a route update.

Reply:
This is to help filters need to pre-process per_route_config.  per_route_config can be dynamically changed anytime.  But you don't want to pre-process the config for each request.  This hash will help to detect config change, and skip pre_processing step:

This is how it can be used,  For each request in decodeHeader()

    auto hash = route()->perFilterConfigHash("my_filter_name");
    if (hash == 0) // handle not per_route_config case

     obj =  global_obj_map.find(hash);
     if (obj == null) {
             obj = pre_process_config( route()->perFilterConfig("my_filter_name") );
             global_obj_map[hash] = obj;
     }

This will also achieve:  different routes with same config can share the same pre_processed object.

Reply:
This is not how I would recommend implementing this. (The above is not thread safe.). I have a pretty good idea of how to efficiently implement this and give you the functionality you need. Let's do this in a follow up.

Reply:
In istio mixer-client filter,  the "global_map" actually is per_thread_global_map.  so this approach works for us. 
I agreed that the filter_factory callback approach is better, the global_map can be used by all worker threads. Let us do this in a followup PR.
Istio is waiting on this feature.   If you don't have anybody to work on this, I can help.